man/dbdcnv.1
.TH DBDCNV 1 "12 September 1995"
.FB
.SH NAME
dbdcnv \- define a \fIFirstBase\fP database conversion dictionary
.SH SYNOPSIS
.B dbdcnv
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
.SH DESCRIPTION
.I Dbdcnv
allows definition and modification of a 
.I FirstBase
database conversion dictionary. This dictionary will be used
by 
\fIdbcgen\fP(1)
to generate converted databases.
.PP
A database conversion dictionary consists of 
a series of field descriptors. Each field descriptor 
contains at least a 
.I field_name,
with an optional
.I field_size.
.PP
\fIDbdcnv\fP(1)
allows creation and deletion of conversion dictionary
field descriptors, as well as interactive
information on the status of each field in the conversion.
Every conversion field falls into one of the following three categories:
.PD 1
.TP 5
.I existing field_name - matching sizes
This field exists in the current database and the request size matches
the actual size. This field will be copied as is to the converted
database, and is marked
.B <image>
in the conversion status column.
.TP 5
.I existing field_name - mis-matched sizes
This field exists in the current database and the request size does not
match the actual size. Truncation or expansion of any field is allowed.
(Except a date field!). This field will be marked with either
.B <truncated>
or 
.B <expanded>
in the conversion status column.
.TP 5
.I new field_name
This field does not exist in the current database.
This new name signifies that a new field of the specified size
will be added to every record in the database.
The contents of the field will be blanks.
.PD
.PP
The status of every conversion field is displayed with the field
on the screen during the
editing session. 
Fields not listed will not appear in the new database.
It is also possible to specify the name of the
new database-to-be.
.PP
For each pre-existing conversion field entered, the default size of the
field is the size of that field as defined by the database dictionary.
Thus, by using a series of <RETURN>s, \fIdbdcnv\fP(1) will set up an image
conversion dictionary.
.PP
Like all other \fIFirstBase\fP dictionary editors, \fIdbdcnv\fP(1) has a
set of commands
that allow screen oriented modification of the conversion dictionary.
The following is a summary of the commands accepted by 
\fIdbdcnv\fP (1):
.PD 0
.TP 7
.B <CTL>-H
displays a summary of these commands.
.TP 7
.B ?[dfhiN]
displays help on the database, formulas,
user defined help files, index, or field N.
.TP 7
.B <RETURN>
(return, or enter) pages forward one page in the database
dictionary (same as 'f').
Wrapping occurs at end of file.
.TP 7
.B @
go into auto add mode.
.TP 7
.B #
allow changes to line number #.
.TP 7
.B -
standard END key \(em exits program and keeps all changes.
.TP 7
.B <CTL>-X
aborts the program. The effect is to exit without
overwriting the file, meaning all changes are lost.
.TP 7
.B b
pages backward one page in the database conversion dictionary. 
Wrapping occurs at beginning of file.
.TP 7
.B d
delete function. Allows deletion of a single dictionary entry.
All entries below deleted entry slide up one slot.
.TP 7
.B i
insert function. Inserts a standard dictionary entry before any given
field.
All entries below inserted entry slide down one slot.
.TP 7
.B m
miscellaneous information screen. Changes can be made to the conversion
output file name.
The default is 'convrt'.
This screen is jumped to on the way out of \fIdbdcnv\fP(1) automatically.
.PD
.PP
A list of simple editor commands appears on the 24th line of the 
screen at all times, with help available by using the standard help 
signal (<CTL>-H).
Complete interactive error checking is done on the contents of
all conversion descriptors.
.PP
Upon successful completion of a conversion dictionary, the user is given a
chance to run the conversion generator. All arguments given
to \fIdbdcnv\fP(1) are passed on to \fIdbcgen\fP(1).
The processes are overlaid,
meaning no new shells are generated in using this option.
.SH FILES
.PD 0
.TP 10
dbase
default name for the database
.TP 10
index
default name for the index
.TP 10
convrt
default name for the converted database
.TP 10
*.idictc
dictionary for conversion generator.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.PD
.SH SEE ALSO
dbedit(1), dbcgen(1), input(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbdcnv\fP(1)
will allow editing and modifying of any conversion dictionary at any time.
These dictionary files should be write protected for complete protection.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbpgen/dbdcnv.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbdcnv.c,v 9.1 2001/02/16 19:44:55 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbdcnv.c	8.2 05 Apr 1996 FB";
#endif

/* 
 *  dbdcnv.c - screen editor for conversion (idictc) file
 */

#include <fb.h>
#include <fb_vars.h>
#include <dbd.h>
#include <pgen.h>

#define HLP_DBDCNV	"dbdcnv.hlp"

extern short int cdb_locklevel;
extern short int cdb_askgen;
extern char *cdb_CGEN;

static int col[8] = {1, 6, 20, 35, 50, 60, 70};

static int dot = 0;			/* current fb_field */
static int lastitem = 0;		/* last list item */

static fb_field *itemlist[FB_MAXKEEP] = { NULL };
static int csize[FB_MAXKEEP] = { 0 };
static char fname[FB_MAXNAME] = {""};	/* output file name */

fb_database *hp;

#define SHOWP 3				/* ending of shown screen columns */

static edit_idictc();
static fillscreen();
static escreen();
static void fdelete();
static void finsert();
static fadder();
static eitem();
static gettitle();
static changetitle();
static fput();
static newdictc();
static read_idictp();
static headers();

/* 
 * dbdcnv - main driver for dbdcnv
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
      
      {
         int st, i;
	 char idictc[FB_MAXNAME];

	 cdb_locklevel = -1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 fb_initdbd(hp);
         if (fb_getd_dict(hp) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, hp->ddict, NIL);
         else
            for (i = 0; i < (hp->nfields); i++)
               fb_nounders(cdb_keymap[i]);
	 fb_scrhdr(hp, "Define Conversion");
         headers();
	 sprintf(idictc, "%sc", hp->idict);
         st = edit_idictc(idictc);
         if (st == FB_END){
	    gettitle();
            newdictc(idictc);
	    if (cdb_askgen)
	       fb_chainto("Generate Conversion", cdb_CGEN, argv);
	    }
         fb_ender();
      }

/* 
 *  edit_idictc - edit an index.idictc or allow creation of one.
 */
 
   static edit_idictc(p)
      char *p;
      
      {
         int st, i;
	 FILE *fs;

         fb_basename(fname, DCNV);
         if ((fs = fopen(p, "r")) != NULL)
	    read_idictp(fs);
	 for (i = lastitem + 1; i < FB_MAXKEEP; i++)
	    itemlist[i] = NULL;
         dot = 1;
         fillscreen();
         while ((st = escreen()) != FB_END)
            if (st == FB_ABORT)
               break;
            else if (st >= dot && st <= dot + 9 && st <= lastitem){
               fb_cx_push_env("6D", CX_KEY_SELECT, NIL);
               eitem(st, 0);
               fb_cx_pop_env();
               }
            else if (st >= 1 && st <= lastitem){
               dot = fb_onpage(st);
               fillscreen();
               }
            else
               fb_screrr("Invalid Input");
         return(st);
      }

/* 
 *  fillscreen - fill a screen with dictionary items.
 */
 
   static fillscreen()
      {
         int j, i, p;
         
         fb_move(4, 1); fb_clrtobot();
         j = dot + 9;
         for (i = dot; i <= j && i <= lastitem; i++)
            for (p = 0; p <= SHOWP; ++p)
               fput(i, p);
	 fb_infoline();
         fb_cx_set_viewpage(dot);
      }
            
/* 
 *  escreen - set up a screen, accept commands for screen control.
 */
 
   static escreen()
      {
         int st;
	 char com[6];
         
         for (;;){
            fb_cx_push_env("E1", CX_DBD_SELECT, NIL);
            if (lastitem > 10)
               fb_cx_add_buttons("Ge");
            fb_cx_add_buttons("XH");
            st = fb_anychange(com);
            fb_cx_pop_env();
            switch (st) {
               case FB_ABORT:
	          if (fb_mustbe('y', "Really Quit? (y/n) ",
                        cdb_t_lines, 1) == FB_AOK){
                     fb_screrr("Conversion Dictionary File Unchanged");
		     return(st);
		     }
		  break;
               case ADDMODE:
                  fadder();
                  break;
               case DELMODE:
	          if (lastitem > 0)
                     fdelete();
		  else
		     continue;
                  break;
               case FB_END:
                  return(st);
	       case HELPMODE:
	          fb_move(2, 1); fb_clrtobot();
		  fb_help(com, hp);
		  fb_move(4,1); fb_clrtobot();
		  headers();
		  break;
	       case INSERTMODE:
	          if (lastitem > 0 && lastitem < FB_MAXKEEP)
	             finsert();
		  else
		     continue;
		  break;
               case PAGEF:
                  dot = dot + 10;
                  if (dot > lastitem)
                     dot = 1;
                  break;
               case PAGEB:
                  dot = dot - 10;
                  if (dot < 1)
                     dot = fb_onpage(lastitem);
                  break;
	       case FB_QHELP: 
	          fb_move(2, 1); fb_clrtobot();
	          fb_fhelp(HLP_DBDCNV);
		  headers();
		  break;
	       case MINFOMODE:
	          gettitle();
		  headers();
		  break;
	       case FB_AOK:
                  return(atoi(com));
               default:
                  return(st);
               }
            fillscreen();
            }
      }

/* 
 *  fdelete - fb_delete a dictionary item 
 */
 
   static void fdelete()
      {
         int st, n;
         
         fb_move(cdb_t_lines, 1);
         fb_clrtoeol();
         fb_printw("Delete what number ? (#,-) ");
         st = fb_input(cdb_t_lines, 28, 3, 0, FB_INTEGER, (char *) &n,
            FB_ECHO, FB_OKEND, FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > lastitem)
	    return;
	 if (csize[n] < 0)
	    fb_free((char *) itemlist[n]);
         for (; n < lastitem; n++){
            itemlist[n] = itemlist[n + 1];
	    csize[n] = csize[n + 1];
	    }
         lastitem--;
      }

/* 
 *  finsert - insert a defaulted dict item before a given item.
 */
 
   static void finsert()            
      {
         int st, n, rn;

	 fb_fmessage("Insert before what number (#,-) ? ");
         st = fb_input(cdb_t_lines, 36, 3, 0, FB_INTEGER, (char *) &n,
            FB_ECHO,FB_OKEND,FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > lastitem)
	    return;
	 rn = n;
         for (n = lastitem; n >= rn; n--){		/* go backwards */
            itemlist[n+1] = itemlist[n];
	    csize[n+1] = csize[n];
	    }
	 itemlist[rn] = cdb_kp[0];			/* default to first fb_field */
         lastitem++;
      }

/* 
 *  fadder - go into auto add mode.
 */
 
   static fadder()
      {
         int st;

         dot = fb_onpage(lastitem);
	 if ((lastitem + 1) < (dot + 10))
            fillscreen();
         for (;lastitem < FB_MAXKEEP;){
            if (++lastitem >= dot + 10){
               dot += 10;
               fb_move(4, 1); fb_clrtobot(); fb_infoline();
               }
            fb_cx_push_env("6D", CX_KEY_SELECT, NIL);
            st = eitem(lastitem, 1);
            fb_cx_pop_env();
            if (st == FB_END){
	       lastitem--;
               break;
	       }
            }
      }

/* 
 *  eitem - edit an entire item.
 */
 
   static eitem(it, new)
      int it, new;

      {
         int row, st, itemp, i;
         char buf[FB_MAXLINE], *fb_trim(), temp[FB_MAXLINE];
	 fb_field *fp = NULL;

         fb_move(cdb_t_lines, 1); fb_clrtoeol();
         if ((row = (it % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
         if (new == 1)
            fput(it, 0);
	    
	 /*
	  *  get id
	  */

         for (;;){
	    if (new)
               fb_fmessage("Enter FieldName, -=End");
	    else
               fb_fmessage("Enter FieldName, <CTL>-X=Skip");
            st = fb_input(row, col[1], FB_TITLESIZE, 0, FB_ALPHA, 
                 temp, FB_ECHO, FB_OKEND, FB_CONFIRM);
            if (st == FB_END && new == 1){
               fb_move(row, 1); 
               fb_clrtoeol();
               return(FB_END);
               }
            else if (st == FB_END)
               st = FB_DEFAULT;
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1)){
	       if ((it - 1) < hp->nfields)
	          fp = cdb_kp[it - 1];
	       else
	          fp = cdb_kp[0];
	       strcpy(temp, fp->id);
	       fb_move(row, col[1]); fb_printw("%s", temp);
	       st = FB_AOK;
	       }
            if (st == FB_ABORT)
               break;
            else if (st == FB_AOK){
               fb_trim(temp);
	       for (i = 1; i <= lastitem; i++){
	          if (i != it)
		     if (equal(temp, itemlist[i]->id))
		        break;
		  }
	       if (i <= lastitem){
	          fb_screrr("That Field Id already used");
		  continue;
		  }
	       fp = fb_findfield(temp, hp);
	       }
	    if (itemlist[it] != NULL){
	       if (csize[it] < 0)		/* clean out what was there */
	          fb_free((char *) itemlist[it]);
	       itemlist[it] = NULL;
	       csize[it] = 0;
	       }
	    if (fp == NULL){
	       for (i = 0; temp[i]; i++){
	          if ((temp[i] >= 'A' && temp[i] <= 'Z') ||
		      (temp[i] >= 'a' && temp[i] <= 'z') ||
		      (temp[i] >= '0' && temp[i] <= '9') ||
		      (temp[i] == FB_BLANK))
		     continue;
		  else{
		     fb_screrr("FieldName must be Alphanumeric!");
		     break;
		     }
		  }
	       if (temp[i])
	          continue;
	       fb_fmessage("NEW field!");
	       fp = (fb_field *) fb_malloc(sizeof(fb_field));
	       fp->id = (char *) fb_malloc(strlen(temp) + 1);
	       strcpy(fp->id, temp);
	       fp->size = 10;
	       fp->type = 'a';
	       csize[it] = -10;	/* so type will prompt */
	       }
	    else{
	       fb_fmessage("Existing field.");
	       csize[it] = fp->size;
	       }
            itemlist[it] = fp;
	    break;
            }
         fput(it, 1);
	 fput(it, 3);
	 
         /* 
	  *  get type -- only if new, meaning csize < 0
	  */

	 for(;;){
	    if (csize[it] > 0)		/* no chaning types if fb_field exists */
	       break;
	    fb_fmessage("Enter type [aAcdfFnN$] for NEW field");
            st = fb_input(row, col[2], 1, 0, FB_ALPHA, buf, FB_ECHO, FB_NOEND, FB_CONFIRM);
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1))
               buf[0] = 'a';
            else if (st == FB_ABORT)
               break;
            switch(buf[0]){
	       case I_FORMULA:
	       case I_LINK:
               case I_ALPHA: case I_NUMERIC: case I_DATE: case I_DOLLARS:
               case I_STRICTALPHA: case I_FLOAT:
	       case I_CHOICE: case I_POS_NUM: case I_UPPERCASE:
	       case I_CCHOICE:
	       case I_BINARY:
	       case I_EXCHOICE:
                  st = FB_AOK;
                  break;
	       case I_TIME: 
	       case I_YESNO:
	          fb_screrr("Unimplemented Type");
		  st = FB_ERROR;
		  break;
               default:
                  fb_screrr("Invalid Type: use [aAcdfFnNty$]");
                  continue;
               }
            if (st == FB_AOK){
               itemlist[it]->type = buf[0];
               break;
               }
            }
	 fput(it, 2);
	 
        /* 
	 *  size
	 */
	 
         for(;;){
	    if (itemlist[it]->type ==FB_DATE){
	       if (csize[it] < 0)
	          fp->size = csize[it] = -6;	/* new date fb_field */
	       break;
	       }
	    fb_fmessage("Enter Size, <RET>=Default, <CTL>-X=Skip");
            st = fb_input(row, col[3], 6, 0,FB_INTEGER,(char *) &itemp,
	          FB_ECHO,FB_NOEND,FB_CONFIRM);
	    if (st == FB_DEFAULT || st == FB_ABORT){
	       if (csize[it] < 0)
	          itemp = -csize[it];	/* new - make pos so -itemp works */
	       else
	          itemp = csize[it];
	       }
	    else if (itemp < 0)
	       continue;
	    else if (itemp >= FB_NUMLENGTH && 
	          itemlist[it]->type != FB_ALPHA && 
	          itemlist[it]->type != FB_BINARY && 
	          itemlist[it]->type != FB_STRICTALPHA &&
		  itemlist[it]->type != FB_CHOICE){
		fb_screrr("Must be less than 16 to behave numerically");
		continue;
		}
            if (csize[it] < 0)
	       csize[it] = -itemp;
	    else
	       csize[it] = itemp;
            break;
            }
         fput(it, 3);
	 
	 /*
	  * finally return.
	  */
	  
         return(FB_AOK);
      }

/* 
 *  gettitle - get the information on conversion: output file name
 */
 
   static gettitle()
      {
         fb_move(2, 1); fb_clrtobot(); fb_printw("Miscellaneous Information");
	 fb_move(12,5); fb_clrtoeol(); fb_printw("Output File:");
	 fb_move(12,20); fb_printw(FB_FSTRING, fname);
	 while (changetitle() != FB_END)
	    ;
      }

/* 
 *  changetitle - ask about changing title, accept fb_input for change 
 */
 
   static changetitle()
      {
         char temp[FB_MAXLINE], dname[FB_MAXNAME];
	 int st;
	 
         if (fb_mustbe('y', "Any Information Change (y/<RET>)",
               cdb_t_lines, 1) != FB_AOK)
	    return(FB_END);
	 fb_fmessage("Enter value, <CTL>-X=Skip, <RET>=Default");
	 fb_basename(dname, cdb_db->dbase);
	 
	 /*
	  *  get fname 
	  */
	 for (;;){
	    st = fb_input(12,20,59,0,FB_ALPHA,temp, FB_ECHO, FB_NOEND, FB_CONFIRM);
	    if (st != FB_ABORT){
	       if (st == FB_DEFAULT)
		  fb_basename(temp, DCNV);
	       strcpy(fname, fb_trim(temp));
	       }
	    fb_move(12,20); fb_clrtoeol(); fb_printw(FB_FSTRING, fname);
	    if (!equal(fname, dname))
	       break;
	    fb_screrr("Target and source databases must be distinct.");
	    }
	 return(FB_NOEND);		/* not FB_END */
      }

/* 
 *  fput - fb_put an entire item to a constant screen location 
 */
 
   static fput(i, pos)
      int i, pos;
      
      {
         int row, v;
         
         if ((row = (i % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
         fb_move(row, col[pos]);
         switch (pos){
            case 0: fb_printw("%3d)", i); break;
            case 1: fb_printw("%-10s", itemlist[i]->id); break;
	    case 2: fb_printw(FB_FCHAR, itemlist[i]->type); break;
	    case 3:
	       if (csize[i] < 0)
	          v = -csize[i];
	       else
	          v = csize[i];
	       fb_printw("%6d", v);
	       fb_move(row, col[pos+1]); fb_clrtoeol();
	       if (csize[i] < 0)
		  fb_printw("new");
	       else if (csize[i] > itemlist[i]->size)
		  fb_printw("expanded");
	       else if (csize[i] < itemlist[i]->size)
		  fb_printw("truncated");
	       else
	          fb_printw("<image>");
            }
      }

/* 
 *  newdictc - printout a new dictionary list of itemlist 
 */
 
   static newdictc(p)
      char *p;
      
      {
         int i;
	 fb_field *fp;
	 FILE *fs, *fb_mustfopen();
	 
	 fs = fb_mustfopen(p, "w");
	 for (i = 1; i <= lastitem; i++){
	    fp = itemlist[i];
	    fb_underscore(fp->id, 1);
	    fprintf(fs, "%s %c %d\n", fp->id, fp->type, abs(csize[i]));
	    }
	 fprintf(fs, "%%\n");
	 if (strlen(fb_trim(fname)) > 0)
	    fprintf(fs, "%s\n", fname);
	 fclose(fs);
      }
      
/* 
 *  read_idictp - read in an existing idctp file for editing 
 */
 
   static read_idictp(fs)
      FILE *fs;
      
      {
         char line[FB_MAXLINE], word[FB_MAXLINE];
	 fb_field  *fp;
	 int p, q, v;
	 
	 q = 0;
	 while (fgets(line, FB_MAXLINE, fs) != NULL){
	    if (line[0] == '%')
	       break;
	    q++;
	    p = fb_getword(line, 1, word);
	    fb_underscore(word, 0);
	    if ((fp = fb_findfield(word, hp)) != NULL){
	       itemlist[q] = fp;
	       csize[q] = fp->size;
	       }
	    else {
	       fp = (fb_field *) fb_malloc(sizeof(fb_field));
	       fp->id = (char *) fb_malloc(strlen(word) + 1);
	       fp->size = 10;
	       fp->type = 'a';
	       strcpy(fp->id, word);
	       itemlist[q] = fp;
	       csize[q] = -10;
	       }
	    p = fb_getword(line, p, word);
	    if (csize[q] < 0)
	       itemlist[q]->type = word[0];
	    p = fb_getword(line, p, word);
	    v = atoi(word);
	    if (csize[q] < 0)
	       csize[q] = -v;
	    else
	       csize[q] = v;
	    }
	 lastitem = q;
	 if (line[0] != '%')
	    fb_xerror(FB_BAD_DICT, hp->idict, "no terminator");
         fb_basename(fname, DCNV);
	 if (fgets(line, FB_MAXLINE, fs) != NULL){
	    line[strlen(line)-1] = NULL;
	    strcpy(fname, fb_trim(line));
	    }
      }

/* 
 * headers - print header at top of screen 
 */
 
   static headers()
      {
         fb_move(2, 1); fb_clrtoeol();
	 fb_move(3, 1); fb_clrtoeol();
         fb_move(2,col[1]);   fb_printw("Field");
         fb_move(3,col[1]);   fb_printw("=====");
	 fb_move(2,col[2]-1); fb_printw("Type");
	 fb_move(3,col[2]-1); fb_printw("====");
	 fb_move(2,col[3]+2); fb_printw("Size");
	 fb_move(3,col[3]+2); fb_printw("====");
         fb_move(2,col[4]-3); fb_printw("Conversion Status");
         fb_move(3,col[4]-3); fb_printw("=================");
      }
      
/* 
 *  usage message 
 */
   usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbdcnv [-d dbase] [-i index]", NIL);
      }
